src/Mutation.tsx

```diff
@@ -0,0 +1,209 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import ApolloClient, {
+  PureQueryOptions,
+  MutationUpdaterFn,
+  ApolloError,
+} from 'apollo-client';
+const invariant = require('invariant');
+import { DocumentNode } from 'graphql';
+const shallowEqual = require('fbjs/lib/shallowEqual');
+
+import { OperationVariables } from './types';
+import { parser, DocumentType } from './parser';
+
+export interface MutationResult<TData = any> {
+  data: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export interface MutationProps<TData = any, TVariables = OperationVariables> {
+  mutation: DocumentNode;
+  variables?: TVariables;
+  optimisticResponse?: Object;
+  refetchQueries?: string[] | PureQueryOptions[];
+  update?: MutationUpdaterFn;
+  children: (
+    mutateFn: () => void,
+    result?: MutationResult<TData>,
+  ) => React.ReactNode;
+  onCompleted?: (data: TData) => void;
+  onError?: (error: ApolloError) => void;
+}
+
+export interface MutationState<TData = any> {
+  notCalled: boolean;
+  error?: ApolloError;
+  data?: TData;
+  loading?: boolean;
+}
+
+const initialState = {
+  notCalled: true,
+};
+
+class Mutation<
+  TData = any,
+  TVariables = OperationVariables
+> extends React.Component<
+  MutationProps<TData, TVariables>,
+  MutationState<TData>
+> {
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+  
+  static propTypes = {
+    mutation: PropTypes.object.isRequired,
+    variables: PropTypes.object,
+    optimisticResponse: PropTypes.object,
+    refetchQueries: PropTypes.oneOfType([
+      PropTypes.arrayOf(PropTypes.string),
+      PropTypes.arrayOf(PropTypes.object)
+    ]),
+    update: PropTypes.func,
+    children: PropTypes.func.isRequired,
+    onCompleted: PropTypes.func,
+    onError: PropTypes.func
+  };
+
+  private client: ApolloClient<any>;
+
+  constructor(props: MutationProps<TData, TVariables>, context: any) {
+    super(props, context);
+
+    this.verifyContext(context);
+    this.client = context.client;
+
+    this.verifyDocumentIsMutation(props.mutation);
+
+    this.state = initialState;
+  }
+
+  componentWillReceiveProps(nextProps, nextContext) {
+    if (
+      shallowEqual(this.props, nextProps) &&
+      this.client === nextContext.client
+    ) {
+      return;
+    }
+
+    this.verifyDocumentIsMutation(nextProps.mutation);
```

Comment:
Do we want to do this only if the mutation is different from the previous one? It involves calling `parser` which I think does a couple things that may add unnecessary overhead if the mutation doesn't change, which should be most of the time

Reply:
Nice! good tip